[Feature] add kube-router (alternative CNI-Plugin that can also replace kube-proxy)

iwilltry42 opened the pull request:
This PR adds the template source for the `kube-router` CNI-Plugin, which can also run a service proxy replacing kube-proxy.

Related PRs:
- ~https://github.com/rancher/types/pull/1058~ (dropped, because types are now included in rancher/rke)
- https://github.com/rancher/rke/pull/1831

Commit:
add rke template for kube-router cni

Commit:
replace server value with var

Commit:
add template mapping

Commit:
use apiroot

Commit:
add dnspolicy

Commit:
add kuberoutercni image

Commit:
update cni-plugins

Commit:
add dnsConfig for kube-router

Commit:
enable kube-router hostport

Commit:
add metrics port and ability to disable the firewall

Commit:
update vendor

Commit:
only download cni if bin dir is empty

Commit:
remove replace stanza in go.mod

Commit:
Merge branch 'master' of https://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

iwilltry42 committed:
Merge branch 'master' of https://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

Oats87 committed:
Add ability to set flexvol directory for Canal and Calico manifests

Oats87 committed:
Base copy

Oats87 committed:
Bump Kubernetes Versions -- v1.14.10-rancher1-1, v1.15.10-rancher1-1, v1.16.7-rancher1-1, v1.17.3-rancher1-1

Oats87 committed:
Generate data

Oats87 committed:
Merge pull request #139 from Oats87/new-k8s-feb-18-2020

New K8s versions v1.14.10, v1.15.10, v1.16.7, v1.17.3

Oats87 committed:
Generate data

Oats87 committed:
Merge pull request #144 from Oats87/dev-v2.3-flexvol

[v2.3] Add custom FlexVolPluginDir capabilities to Calico and Canal

Commit:
Merge pull request #146 from rancher/dev-v2.3

k8s 1.14.10, v1.15.10, v1.16.7, v1.17.3

superseb committed:
base copy

superseb committed:
Add k8s 1.15.11, 1.16.8, 1.17.4

superseb committed:
generated data

superseb committed:
Merge pull request #156 from superseb/v23_k8s_versions_20201203

[dev-v2.3] Add k8s 1.15.11, 1.16.8, 1.17.4

superseb committed:
Add etcd options for new k8s versions

- For v1.15.11, v1.16.8

superseb committed:
generated data

superseb committed:
Merge pull request #164 from superseb/v23_new_k8s_svcoptions

[dev-v2.3] Add etcd options for new k8s versions

superseb committed:
Add Calico/Canal v3.13.0 to new k8s versions

superseb committed:
generated data

superseb committed:
Merge pull request #166 from superseb/v23_calico_v3130_updates

[dev-v2.3] Add Calico/Canal v3.13.0 to new k8s versions

superseb committed:
Merge pull request #168 from rancher/dev-v2.3

Add k8s 1.15.11, 1.16.8, 1.17.4

luthermonson committed:
rke-tools v0.1.54

luthermonson committed:
generated

luthermonson committed:
Merge pull request #173 from luthermonson/dev-v2.3

rke v0.1.54 azure gov cloud fix

luthermonson committed:
default to v1.17.4-rancher1-2

luthermonson committed:
Merge pull request #175 from luthermonson/dev-v2.3

setting default to v1.17.4-rancher1-2

luthermonson committed:
rke-tools:v0.1.55

luthermonson committed:
generated

Commit:
Merge pull request #181 from luthermonson/tools-55

version bump rke-tools:v0.1.55

luthermonson committed:
rke-tools:v0.1.56

luthermonson committed:
generated

Commit:
Merge pull request #182 from luthermonson/tools56

rke-tools:v0.1.56

kinarashah committed:
v1.15.11-rancher1-0

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #185 from kinarashah/kt

v1.15.11-rancher1-0

kinarashah committed:
rancher min version 2.3.6: v1.15.11-rancher1-2, v1.16.8-rancher1-2, v1.17.4-rancher1-2

kinarashah committed:
generated data

Commit:
Merge pull request #188 from kinarashah/mink

[2.3] rancher min version 2.3.6: v1.15.11-rancher1-2, v1.16.8-rancher1-2, v1.17.4-rancher1-2

Commit:
Add comments for saving template rancher1-1 for k8s 1.15

kinarashah committed:
rancher min version 2.3.5: v1.15.11-rancher1-2, v1.16.8-rancher1-2, v1.17.4-rancher1-2

kinarashah committed:
generated data

Commit:
Merge pull request #193 from kinarashah/mk

change min rancher version to 2.3.5

Commit:
Merge pull request #194 from rancher/dev-v2.3

Release v2.3.6

iwilltry42 committed:
Merge branch 'master' of ssh://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

superseb committed:
Set new min versions for introduced k8s versions

Rancher v2.3.2 and lower cannot handle template functions, so we need to exclude this version as it contains a CNI template for Calico and Canal that contains a template function

kinarashah committed:
1.15.11-rancher1-0 and 1.14.10-rancher1-0 changes from 2.2.11

kinarashah committed:
generated data

Commit:
Merge pull request #198 from kinarashah/v2

[dev-v2.3] 1.15.11-rancher1-0 and 1.14.10-rancher1-0 changes from 2.2.11

superseb committed:
Restrict new CNI templates for v2.3+ k8s versions

superseb committed:
generated data

Commit:
Merge pull request #199 from superseb/v23_canalcalico_restrictions_v22

[dev-v2.3] Restrict new CNI templates for v2.3+ k8s versions

kinarashah committed:
add calicoControllers to system images for out of band k8s version 1.14.10-rancher1-0

For out of band/KDM k8s versions, RKE validates against empty system images
even if it is not getting actually used by network template.

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #202 from kinarashah/clc

[dev-v2.3] add calicoControllers to system images for out of band k8s version 1.14.10-rancher1-0

Commit:
Merge pull request #205 from rancher/dev-v2.3

[2.3] k8s 1.14.10-rancher1-0 and k8s 1.15.11-rancher1-0

superseb committed:
generated data

superseb committed:
Add nodelocal DNS

superseb committed:
Add new k8s versions with images for nodelocal DNS

superseb committed:
vendor types

superseb committed:
generated data

superseb committed:
Add missing Calico controllers images

superseb committed:
generated data

mrajashree committed:
Base copy

mrajashree committed:
1.16 and 1.17 patch versions

mrajashree committed:
Generated data changes

superseb committed:
Use nodelocal template for all k8s versions

superseb committed:
generated data

superseb committed:
Add support for version checking for templates

mrajashree committed:
Include nodelocal and k8s patch in one version

1.17.4-rancher1-3 and 1.16.8-rancher1-3 were added for nodeLocal DNS for 2.3.7
1.17.5-rancher1-1 and 1.16.9-rancher1-1 were added for k8s patch releases for 2.3.7
Only one of them will be presented during cluster creation, 1.17.5 and 1.16.9
So this commit combines 1.17.4-rancher1-3 and 1.17.5-rancher1-1 into a single version
And the same for 1.16.8-rancher1-3 and 1.16.9-rancher1-1
Also sets the rancher min version to 2.3.7-rc0 for these versions so they show up
2.3.7-rc1 onwards

mrajashree committed:
Generated data changes

mrajashree committed:
Change rancherMinVersion for 1.15 nodelocal

mrajashree committed:
Generated data changes

Commit:
Merge pull request #226 from rancher/dev-v2.3

v2.3.7

ryansann committed:
fix merge conflicts from cherry-pick

ryansann committed:
fix merge conflicts from cherry-pick of 2f5527f8ff85dc4c70617f7a88da0059e7a00c31

ryansann committed:
remove rc0 from v1.17.6-rancher1-1

ryansann committed:
generated data

ryansann committed:
generated data

ryansann committed:
add missing template mapping for nginx ingress

ryansann committed:
fix merge conflict from cherry pick of db39588315adda62eb502994f23a3c9d1aaa17ef

ryansann committed:
Merge pull request #247 from ryansann/ingress-upgrade-2.3

[2.3] Backport ingress upgrade & K8S v1.15.12, v1.16.10, v1.17.6

Commit:
Update to flannel-cni:v0.3.0-rancher6

Commit:
Updated to use the tags with Windows 1909 support in hyperkube images

Commit:
Update calico to v3.13.4

Commit:
Update default to v1.17.6-rancher2-1

Commit:
weave 2.6.3

Commit:
go generate

kinarashah committed:
add v1.15.12-rancher1-1

superseb committed:
Add Calico v3.13.4 template for new k8s versions

superseb committed:
generated data

Commit:
Merge pull request #258 from deniseschannon/cni-updates-dev-v2.3

[dev-v2.3] flannel-cni:v0.3.0-rancher6, calico:v3.13.4, updated hyperkube images due to windows 1909 builds

Commit:
Merge pull request #266 from superseb/v23_calicov3134_templates

[dev-v2.3]  Add Calico v3.13.4 template for new k8s versions

ryansann committed:
add v1.15.12 service options and remove UpdateStrategy from ingress template

ryansann committed:
generated data

Commit:
Merge pull request #270 from ryansann/nginx-template-fix-2.3

[2.3] Fix nginx template for 2.3.x

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #271 from kinarashah/dv23

[2.3] add v1.15.12-rancher1-1

kinarashah committed:
set updateStrategy on ingress for upgraded 2.2.x clusters

ingress daemonset in rke 0.2.10 had apiVersion=extensions/v1beta1
which defaults to onDelete strategy. We change apiVersion in new
template to apps/v1, but without explicitly changing the strategy,
it stays onDelete.

kinarashah committed:
generated data

kinarashah committed:
weave update to 2.6.4

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #274 from kinarashah/dv23

[2.3] set updateStrategy on ingress for upgraded 2.2.x clusters

kinarashah committed:
update min versions for v1.15.12, v1.16.10, v1.17.6

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #277 from kinarashah/dv23

[dev-v2.3] update min version info

kinarashah committed:
calico/canal template key change for v1.15.12-rancher1-1

kinarashah committed:
generated data

kinarashah committed:
introduce v1.15.12-rancher1-2 for NodeLocal

kinarashah committed:
add max version info for v1.15.12-rancher1

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #279 from kinarashah/tests

[dev-v2.3] changes for v1.15.12-rancher1-x

kinarashah committed:
remove v1.15.12-rancher1-2

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #281 from kinarashah/tests

[dev-v2.3] remove v1.15.12-rancher1-2

Commit:
Merge pull request #268 from rancher/dev-v2.3

[v2.3] Bump in KDM

prachidamle committed:
Change default CIS benchmark version

ryansann committed:
cherry pick 687923cb1052c5e504521d453f84b665218ea4b0, 37faa211bca84e016bc264a9568ce5139e4d70a2, and 885343b9662faeb4d95e65b1933f60933c607c61 / remove 1.18.x references / apply windows and flannel tags to 1.16.10, 1.17.6

ryansann committed:
generated data

ryansann committed:
Merge pull request #293 from ryansann/sec-448-backport-2.3

[2.3] Backport new k8s versions / image tags for flannel and windows

ryansann committed:
fix min versions for v1.15.12-rancher2-3

ryansann committed:
generated data

ryansann committed:
Merge pull request #298 from ryansann/issue/27577

[2.3] Fix version info for v1.15.12-rancher2-3

ryansann committed:
address issues in dev-v2.3 branch

ryansann committed:
generated data

ryansann committed:
v1.16.11-rancher1-1 and v1.17.7-rancher1-1

ryansann committed:
generated data

ryansann committed:
Merge pull request #301 from ryansann/v2.3-fixes

[2.3] fixes / new k8s versions v1.16.11, v1.17.7

ryansann committed:
comment update for version tracking

kinarashah committed:
Revert "v1.16.11-rancher1-1 and v1.17.7-rancher1-1"

This reverts commit 44c69fb48b582dd788d9a00ef4fb96126c6e8ba8.

kinarashah committed:
generated data

kinarashah committed:
Merge pull request #308 from kinarashah/dv23

[dev-v2.3] revert k8s version 1.16.11, 1.17.7

cbron committed:
Merge pull request #300 from rancher/dev-v2.3

2.3.x Updates

Commit:
Bump k3s versions to v1.17.9 and v1.18.6
* Updates channels.yaml so that we now point to v1.17.9+k3s1 and v1.18.6+k3s1 k3s versions.

Signed-off-by: David Nuzik <david.nuzik@rancher.com>

kinarashah committed:
base images

kinarashah committed:
v1.16.13-rancher1-1 and v1.17.9-rancher1-1 for v2.3.8+

kinarashah committed:
generated data

kinarashah committed:
v1.16.13, v1.17.9, v1.18.6

kinarashah committed:
go generate

kinarashah committed:
Merge pull request #320 from kinarashah/dv23

[dev-v2.3] v1.16.13-rancher1-1 and v1.17.9-rancher1-1 for v2.3.8+

kinarashah committed:
Merge pull request #321 from kinarashah/dv24

[dev-v2.4] v1.16.13, v1.17.9, v1.18.6

kinarashah committed:
update min version for v1.16.10-rancher2-2, v1.17.6-rancher2-2

Since we have v1.16.13 with min v2.4.4, it seems inconsistent to have
previous version at 2.4.5. It won't change anything for v2.4.4 because
v1.16.13 is now the latest v1.16 patch version. Same for v1.17.x.

kinarashah committed:
generated data

Commit:
Merge pull request #323 from rancher/dev-v2.3

[v2.3] v1.16.13, v1.17.9

Commit:
go generate

Signed-off-by: David Nuzik <david.nuzik@rancher.com>

davidnuzik committed:
Merge pull request #324 from davidnuzik/update-k3s-july-dev-v2.4-backport

[dev-v2.4] Bump k3s versions to v1.17.9 and v1.18.6

kinarashah committed:
Merge pull request #322 from kinarashah/dv24

[dev-v2.4] update min version for v1.16.10-rancher2-2

prachidamle committed:
Merge pull request #311 from prachidamle/kdm_cis_change

[2.4] Change default CIS benchmark version to CIS 1.5

Commit:
Copy templates

Commit:
Bump rke-tools, set new default, add service option

Bump rke-tools to version v0.1.60 for templates:
- v1.18.6-rancher1-2
- v1.17.9-rancher1-2
- v1.16.13-rancher1-2
- v1.15.12-rancher2-5
Set v1.18.6-rancher1-2 as default
Add service option for:
- v1.16.13-rancher1-2
- v1.15.12-rancher2-5

Commit:
go generate

Commit:
Merge pull request #328 from PennyScissors/dev-v2.4

Bump rke-tools version to v0.1.60

ryansann committed:
bump rke-tools to v0.1.61

ryansann committed:
generated data

ryansann committed:
Merge pull request #331 from ryansann/rke-tools-update-2.4

Bump rke-tools to v0.1.61, fix min versions for v1.16.13, v1.17.9, and v1.18.6

superseb committed:
bump rke-tools to v0.1.62

superseb committed:
generated data

superseb committed:
Merge pull request #335 from superseb/v24_rketools0162

[dev-v2.4] bump rke-tools to v0.1.62

superseb committed:
Bump rke-tools to v0.1.63

superseb committed:
generated data

superseb committed:
Merge pull request #339 from superseb/v24_rketoolsv0163

[dev-v2.4] Bump rke-tools to v0.1.63

luthermonson committed:
version bump rke-tools

luthermonson committed:
generate

luthermonson committed:
Merge pull request #340 from luthermonson/win-prefix-path

Win prefix path

Commit:
K3S - Update k3s v1.17 and v1.18 versions
* Updates v1.17.9 to v1.17.11 and v1.18.6 to v1.18.8
* Note, v1.17.10 and v1.18.7 were skipped as upstream was having issues with these releases and taged new versions.

Commit:
go generate

davidnuzik committed:
Merge pull request #343 from davidnuzik/backport-update-k3s-august-for-realsies

[dev-v2.4] Update k3s v1.17 and v1.18 versions

Commit:
Copy templates

Commit:
Update v1.18.8, v1.17.11, v1.16.14, v1.16.15 and correct image sources

Commit:
Go generate

Commit:
Merge pull request #348 from PennyScissors/update-k8s

[dev-v2.4] Update v1.18.8, v1.17.11, v1.16.14, v1.16.15 and correct image sources

kinarashah committed:
update comment for v1.16.14, v1.16.15, v1.17.11, v1.18.8 - post v2.4.8

Commit:
Copy templates

Commit:
Add 1.17.12, 1.18.9

Commit:
Go generate

Commit:
Merge pull request #362 from PennyScissors/update-k8s-2.4

[dev-v2.4] Update k8s 1.17.12, 1.18.9 + 1.16.15 ingress

Commit:
bump k3s version v1.17.12 v1.18.9 v1.19.2

Commit:
go generate

davidnuzik committed:
Merge pull request #358 from MonzElmasry/backport_k3s_update

[dev-2.4] Update k3s  v1.17.12, v1.18.9 and add v1.19.2

superseb committed:
Add v1.15.12-rancher2-6 and rke-tools v0.1.65

superseb committed:
go generate

superseb committed:
Merge pull request #368 from superseb/v24_rketoolsupdates

[dev-v2.4] Add v1.15.12-rancher2-6 and rke-tools v0.1.65

iwilltry42 committed:
Merge branch 'dev-v2.4' of ssh://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

iwilltry42 committed:
drop changes to calico and canal

iwilltry42 committed:
Merge branch 'dev-v2.5' of ssh://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

iwilltry42 committed:
drop unknown changes in system images

iwilltry42 committed:
add kube router range

ryansann committed:
Add v1.17.13-rancher1-1 and v1.16.15-rancher1-2 k8s versions

ryansann committed:
generated data

ryansann committed:
Merge pull request #377 from ryansann/2.3-oct-14

[dev-v2.3] Backport v1.16.15-rancher1-2 and add v1.17.13-rancher1-1

Commit:
Merge pull request #385 from rancher/dev-v2.3

v1.16.15-rancher1-2 and v1.17.13-rancher1-1

superseb committed:
base copy

superseb committed:
add 1.17.14

superseb committed:
generated data

Commit:
Merge pull request #419 from superseb/v23_k8s1113

[dev-v2.3] add 1.17.14

superseb committed:
update rke-tools to 0.1.66

superseb committed:
generated data

superseb committed:
Merge pull request #422 from superseb/v23_k8s1113_2

[dev-v2.3] update rke-tools to 0.1.66

superseb committed:
update to nginx 0.35 rancher2

superseb committed:
generated data

superseb committed:
Merge pull request #423 from superseb/v23_k8s1113_3

[dev-v2.3] update to nginx 0.35 rancher2

superseb committed:
Merge pull request #424 from rancher/dev-v2.3

add 1.17.14

Commit:
Add v1.17.16-rancher1-1

Commit:
Generated data

superseb committed:
Merge pull request #445 from PennyScissors/devv23-k8s-v11716

[dev-v2.3] Add kdm template v1.17.16-rancher1-1

kinarashah committed:
Merge pull request #451 from rancher/dev-v2.3

[master] Release kdm k8s v1.17.16-rancher1-1

ryansann committed:
v1.17.17-rancher1-1 k8s version

ryansann committed:
generated data

ryansann committed:
Merge pull request #464 from ryansann/01192021-2.3

[dev-v2.3] Add v1.17.17-rancher1-1

kinarashah committed:
Merge pull request #476 from rancher/dev-v2.3

[2.3] Release

iwilltry42 committed:
Merge branch 'master' of ssh://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

superseb commented on rke/k8s_rke_system_images.go:
```diff
@@ -2307,6 +2307,7 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
 			CanalFlexVol:              m("quay.io/calico/pod2daemon-flexvol:v3.8.1"),
 			WeaveNode:                 m("weaveworks/weave-kube:2.5.2"),
 			WeaveCNI:                  m("weaveworks/weave-npc:2.5.2"),
+			KubeRouterCNI:             m("cloudnativelabs/kube-router:v1.0.0"),
```
Can we update this to latest (1.2.2)?

iwilltry42 replied:
:heavy_check_mark:

superseb commented on rke/templates/kube-router.go:
```diff
@@ -0,0 +1,479 @@
+          fi;
+          mkdir -p /opt/cni/bin;
+          if [ ! "$(ls -A /opt/cni/bin)" ]; then
+            wget https://github.com/containernetworking/plugins/releases/download/v0.8.3/cni-plugins-linux-amd64-v0.8.3.tgz -O /tmp/cni-plugins-linux-amd64-v0.8.3.tgz && tar -xf /tmp/cni-plugins-linux-amd64-v0.8.3.tgz -C /opt/cni/bin/;
```
This is provided by rke-tools, why this fallback (which will also break airgap but in general, what is the fallback good for as if binaries are not provided it should just not work)

iwilltry42 replied:
Well.. I didn't know that it was included in rke-tools :thinking: 
Maybe back when I started this PR, it wasn't as I think I copied most things from existing CNI-Plugins. Anyway, if it's included there, I'm going to remove it here :+1:

superseb commented on rke/templates/kube-router.go:
```diff
@@ -0,0 +1,479 @@
+      dnsPolicy: ClusterFirstWithHostNet
+      dnsConfig:
+        nameservers:
+          - 1.1.1.1
```
Possible airgap issue?

iwilltry42 replied:
removed as I updated the whole manifest based on upstream

superseb commented:
* Can you share the upstream template used for this?
* If this goes in, it will be in k8s v1.21, can you update the code for this? (needs to go in `dev-v2.6`)

iwilltry42 committed:
Address PR comments

- update kube-router manifest based on https://raw.githubusercontent.com/cloudnativelabs/kube-router/4e13a1db620c0e3c22989132c52eb2f0c401965f/daemonset/generic-kuberouter-all-features.yaml
- use latest kube-router image v1.2.2 by default
- do not update cni bin in initContainer

Signed-off-by: Thorsten Klein <iwilltry42@gmail.com>

iwilltry42 committed:
Revert "Merge branch 'master' of ssh://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router"

This reverts commit 45e7263bbfbaeadf33850f595d619a1628255abc, reversing
changes made to 1c3a1933ecd5f2fa7f46c29d6b5368fa5b996e71.

iwilltry42 committed:
Merge branch 'dev-v2.6' of ssh://github.com/rancher/kontainer-driver-metadata into feature/add-kube-router

iwilltry42 commented:
@superseb 

>     * Can you share the upstream template used for this?

It's based on https://github.com/cloudnativelabs/kube-router/blob/4e13a1db620c0e3c22989132c52eb2f0c401965f/daemonset/generic-kuberouter-all-features.yaml (updated to this commit just now)
 
>     * If this goes in, it will be in k8s v1.21, can you update the code for this? (needs to go in `dev-v2.6`)

Undid the accidental merge of master, rebased the PR and merged in dev-v2.6 :+1:

iwilltry42 committed:
Merge branch 'dev-v2.6' into feature/add-kube-router
